// @urban-bot/core/dist/components/Poll.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const hooks_1 = require("../hooks/hooks");
const getButtonsByButtonGroup_1 = require("../utils/getButtonsByButtonGroup");
const formatOptionElement_1 = require("../utils/formatOptionElement");
const formatMarkupLanguageElement_1 = require("../utils/formatMarkupLanguageElement");
const getParseMode_1 = require("../utils/getParseMode");
function Poll({ buttons: buttonGroupElement, isNewMessageEveryRender: isNewMessageEveryRenderProp, question, children, isAnonymous, type, withMultipleAnswers, rightOption, explanation, livePeriodSeconds, disableNotification, forceReply, parseMode, replyToMessageId, ...otherProps }) {
    const { $$managerBot, isNewMessageEveryRender: isNewMessageEveryRenderContext, chat, parseMode: parseModeContext, bot, } = hooks_1.useBotContext();
    const finalParseModeQuestion = getParseMode_1.getParseMode(question, parseMode, parseModeContext, bot.defaultParseMode);
    const finalParseModeExplanation = getParseMode_1.getParseMode(explanation, parseMode, parseModeContext, bot.defaultParseMode);
    const finalParseMode = finalParseModeQuestion !== null && finalParseModeQuestion !== void 0 ? finalParseModeQuestion : finalParseModeExplanation;
    const formattedQuestion = formatMarkupLanguageElement_1.formatMarkupLanguageElement(question, finalParseMode);
    const formattedExplanation = formatMarkupLanguageElement_1.formatMarkupLanguageElement(explanation, finalParseMode);
    const formattedButtons = getButtonsByButtonGroup_1.getButtonsByButtonGroup(buttonGroupElement);
    const options = formatOptionElement_1.formatOptionElement(children);
    hooks_1.useAction((ctx) => {
        var _a;
        const { actionId } = ctx;
        const option = options.find(({ id }) => {
            return actionId === id;
        });
        (_a = option === null || option === void 0 ? void 0 : option.onClick) === null || _a === void 0 ? void 0 : _a.call(option, ctx);
    });
    return (react_1.default.createElement("urban-poll", { "$$managerBot": $$managerBot, chat: chat, isNewMessageEveryRender: isNewMessageEveryRenderProp !== null && isNewMessageEveryRenderProp !== void 0 ? isNewMessageEveryRenderProp : isNewMessageEveryRenderContext, data: {
            question: formattedQuestion,
            options,
            isAnonymous,
            type,
            withMultipleAnswers,
            rightOption,
            explanation: formattedExplanation,
            livePeriodSeconds,
            disableNotification,
            replyToMessageId,
            forceReply,
            parseMode: finalParseMode,
            buttons: formattedButtons,
            isReplyButtons: buttonGroupElement === null || buttonGroupElement === void 0 ? void 0 : buttonGroupElement.props.isReplyButtons,
            ...otherProps,
        } }));
}
exports.Poll = Poll;
function Option(_props) {
    return null;
}
exports.Option = Option;
